### PR TITLE
feat: support optimus sensor

### PR DESCRIPTION
**Pull request by irainia**
This PR is to address #400 . Changes related:

* introduces resource manger, with specific to optimus, as a way to communicate to another optimus
* as part of resource manager, additional configuration fields are introduced under `config.yaml` for server
* updates `base_dag.py` to add sensore
* separated dependency resolution, into internal and external, to add more clarity

### CODE REVIEW

File: job/external_dependency_resolver.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+func (e *externalDependencyResolver) FetchStaticExternalDependenciesPerJobName(ctx context.Context, projectID models.ProjectID) (map[string]models.ExternalDependency, []models.UnknownDependency, error) {
+	if ctx == nil {
+		return nil, nil, errors.New("context is nil")
+	}
+	unknownStaticDependenciesPerJobName, err := e.unknownJobDependencyRepository.GetUnknownStaticDependencyNamesPerJobName(ctx, projectID)
+	if err != nil {
+		return nil, nil, err
+	}
+	jobSpecFiltersPerJobName, err := e.toJobSpecFiltersPerJobNameForDependencyNames(unknownStaticDependenciesPerJobName)
```
Comment (sravankorumilli) [inline job/external_dependency_resolver.go:68]: function names are hard to understand especially in the dependency resolution

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: ext/scheduler/airflow2/airflow.go [hunk 2]
```diff
@@ -158,12 +158,12 @@ func (s *scheduler) DeployJobsVerbose(ctx context.Context, namespace models.Name
 	var jobDeploymentDetail models.JobDeploymentDetail
 	for _, result := range runner.Run() {
 		if result.Val != nil {
-			jobDeploymentDetail.Failures = append(jobDeploymentDetail.Failures, result.Val.(models.JobDeploymentFailure))
+			jobDeploymentDetail.DeploymentFailures = append(jobDeploymentDetail.DeploymentFailures, result.Val.(models.JobDeploymentFailure))
```
[the remark this answers is not in the document]
Comment (irainia) [reply ext/scheduler/airflow2/airflow.go:161]: will update

[the remark this answers is not in the document]
Comment (arinda-arif) [reply ext/scheduler/airflow2/airflow.go:161]: will also remove the `DeploymentFailureCount` from the business model but will keep it in the response model.

[the remark this answers is not in the document]
Comment (arinda-arif) [reply ext/scheduler/airflow2/airflow.go]: updated

File: go.mod [hunk 3]
```diff
@@ -66,6 +65,10 @@ require (
 	gorm.io/gorm v1.21.16
 )
 
+require go.uber.org/multierr v1.7.0
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply go.mod]: updated

File: job/dependency_resolver.go [hunk 4]
```diff
@@ -242,34 +248,84 @@ func (*dependencyResolver) resolveHookDependencies(jobSpec models.JobSpec) model
...
+		return nil, nil, err
+	}
+	addToOutput(inferredExternalDependencyPerJobName)
+
+	return output, unknownDependencies, nil
+}
+
+func (d *dependencyResolver) getInternalDependenciesByJobID(ctx context.Context, projectID models.ProjectID) (map[uuid.UUID][]models.JobSpec, error) {
+	output := make(map[uuid.UUID][]models.JobSpec)
+	addToOutput := func(dependencies map[uuid.UUID][]models.JobSpec) {
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/dependency_resolver.go]: updated

File: job/external_dependency_resolver.go (hunk 1 above)
Comment (arinda-arif) [reply job/external_dependency_resolver.go:68]: this will be removed as part of previous feedback.

File: job/external_dependency_resolver.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		unknownJobDependencyRepository: unknownJobDependencyRepository,
+		optimusResourceManagers:        optimusResourceManagers,
+	}, nil
+}
+
+func (e *externalDependencyResolver) FetchInferredExternalDependenciesPerJobName(ctx context.Context, projectID models.ProjectID) (map[string]models.ExternalDependency, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	unknownInferredDependenciesPerJobName, err := e.unknownJobDependencyRepository.GetUnknownInferredDependencyURNsPerJobName(ctx, projectID)
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/external_dependency_resolver.go]: @sravankorumilli updated, please check

File: job/external_dependency_resolver.go (hunk 1 above)
Comment (arinda-arif) [reply job/external_dependency_resolver.go]: removed

File: ext/resourcemgr/optimus_test.go [hunk 6]
```diff
@@ -35,9 +35,9 @@ func (o *OptimusResourceManager) TestGetJobSpecifications() {
 		}
 
 		var ctx context.Context
-		var filter models.JobSpecFilter
+		var unresolvedDependency models.UnresolvedJobDependency
```
Comment (sravankorumilli) [inline ext/resourcemgr/optimus_test.go:38]: would be good if rename could be a separate commit

**Review by sravankorumilli — APPROVED** [review]

File: ext/resourcemgr/optimus_test.go (hunk 6 above)
Comment (arinda-arif) [reply ext/resourcemgr/optimus_test.go:38]: got it, will try to separate commits like this in the future. but for this one i am resolving as we are not only renaming.
